### PR TITLE
New icon/portrait rendering code

### PR DESCRIPTION
**Pull request by giorgiga**
This PR started as a way to push some new code to render unit icons preserving their color and ended up with some extra unrelated change too (see code comments).

Four different ways to blend unit icons with camo/background are implemented; in the image below, the first row is the current MekHQ strategy (take the blue channel and use it to darken the camo) and the third row is the one that is active in this PR (it preserves all pixels that are not perfectly grayscale, and use luminance in lieu of the blue channel in to darken the camo.

In case you are interested, the screenshot is from a window opened by running the main method in `megamek.util.ImagesTest`and the four strategies are implemented in `megamek.util.Images`.

camos

### CODE REVIEW

File: MekHQ/build.gradle [hunk 1]
```diff
@@ -37,7 +37,7 @@ dependencies {
         exclude group: 'org.python', module: 'jython'
         exclude group: 'org.mozilla', module: 'rhino'
     }
-    
+
```
Comment (giorgiga) [inline MekHQ/build.gradle]: Sorry about the whitespace changes – I always forget my editor strips trailing spaces on save

File: MekHQ/src/mekhq/IconPackage.java [hunk 2]
```diff
@@ -163,81 +153,5 @@ public String getStartupScreenImage(int resolutionWidth) {
     public String getLoadingScreenImage(int resolutionWidth) {
         return loadingScreenImages.floorEntry(resolutionWidth).getValue();
     }
-    
-    public static Image buildForceIcon(String category, String filename, DirectoryItems items, LinkedHashMap<String, Vector<String>> iconMap) {
```
Comment (giorgiga) [inline MekHQ/src/mekhq/IconPackage.java]: This has been moved to `megamek.util.Images` (together with the code that processes unit icons and portraits)

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 3]
```diff
@@ -237,7 +242,7 @@
     // we will use the same basic system (borrowed from MegaMek) for tracking
     // all three
     // OK now we have more, parts, personnel, forces, missions, and scenarios.
-    private Map<UUID, Unit> units = new LinkedHashMap<UUID, Unit>();
+    private Map<UUID, Unit> units = new LinkedHashMap<>();
```
Comment (giorgiga) [inline MekHQ/src/mekhq/campaign/Campaign.java]: I couldn't stand those unnecessary type params :-)

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 4]
```diff
@@ -7319,35 +7323,28 @@ public void addLogEntry(Person p, LogEntry entry) {
...
                 }
             }
         }
         return possiblePortraits;
     }
 
     public void assignRandomPortraitFor(Person p) {
-        // first create a list of existing portait strings, so we can check for
-        // duplicates
-        ArrayList<String> existingPortraits = new ArrayList<String>();
```
Comment (giorgiga) [inline MekHQ/src/mekhq/campaign/Campaign.java]: Moved below and made into a Set

File: MekHQ/src/mekhq/campaign/finances/Loan.java [hunk 5]
```diff
@@ -572,7 +572,12 @@ public static String randomRefNumber() {
         }
         return buffer.toString();
     }
-    
+
+    @Override
+    public int hashCode() {
+        return institution.hashCode() + 13 * refNumber.hashCode();
+    }
+
```
Comment (giorgiga) [inline MekHQ/src/mekhq/campaign/finances/Loan.java]: This is not related to images or the other changes – it's just that I noticed this class overrides `equals()` but not `hashCode()`

File: MekHQ/src/mekhq/campaign/force/Force.java [hunk 6]
```diff
@@ -501,93 +430,71 @@ private static void processUnitNodes(Force retVal, Node wn, Version version) {
...
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null) return false;
+        if (getClass() != obj.getClass()) return false;
+        Force other = (Force) obj;
+        return id == other.id
+            && name.equals(other.name)
+            && Objects.equals(parentForce, other.parentForce);
```
Comment (giorgiga) [inline MekHQ/src/mekhq/campaign/force/Force.java]: Previous `equals()` was based on `getFullName()` – this is slightly stricter

File: MekHQ/src/mekhq/campaign/force/Force.java [hunk 7]
```diff
@@ -501,93 +430,71 @@ private static void processUnitNodes(Force retVal, Node wn, Version version) {
...
+        }
+        for (Force sub : subForces) {
+            sub.fixIdReferences(uHash);
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return 123311 * id + name.hashCode();
+    }
```
Comment (giorgiga) [inline MekHQ/src/mekhq/campaign/force/Force.java]: Class implemented `equals()` but not `hashCode()` :-(

File: MekHQ/src/mekhq/campaign/force/ForceStub.java [hunk 8]
```diff
@@ -1,7 +1,7 @@
 /*
  * Force.java
  * 
- * Copyright (c) 2011 Jay Lawson <jaylawson39 at yahoo.com>. All rights reserved.
+ * Copyright (c) 2018 the MegaMek Team.
```
Comment (giorgiga) [inline MekHQ/src/mekhq/campaign/force/ForceStub.java]: So many lines have changed that I guess Jay himself wouldn't recognise this file..

File: MekHQ/src/mekhq/campaign/personnel/Bloodname.java [hunk 9]
```diff
@@ -524,13 +524,18 @@ public Clan() {
 		nameChanges = new ArrayList<DatedRecord>();
 	}
 
+    @Override
+    public int hashCode() {
+        return code.hashCode();
+    }
+
```
Comment (giorgiga) [inline MekHQ/src/mekhq/campaign/personnel/Bloodname.java]: This is not related to images or the other changes – it's just that I noticed this class overrides `equals()` but not `hashCode()`

File: MekHQ/src/mekhq/campaign/personnel/Person.java [hunk 10]
```diff
@@ -1450,10 +1432,10 @@ public void writeToXml(PrintWriter pw1, int indent) {
                         + df.format(recruitment.getTime())
                         + "</recruitment>");
         }
-        for (String skName : skills.keySet()) {
+        skills.keySet().stream().sorted().forEach(skName -> { // sorted to get consistent .cpnx files
```
Comment (giorgiga) [inline MekHQ/src/mekhq/campaign/personnel/Person.java]: Another "opportunity" change :-)

I was diffing savefiles to check the force icon stuff and the changes in order of skills kept popping up as false positives

File: MekHQ/src/mekhq/gui/dialog/ImageChoiceDialog.java [hunk 11]
```diff
@@ -799,20 +798,17 @@ public void setText(String text) {
...
             }
-
             // Try to get the image file.
             try {
                 // Translate the root image directory name.
                 if (Crew.ROOT_PORTRAIT.equals(category))
                     category = ""; //$NON-NLS-1$
                 Image image = (Image) items.getItem(category, name);
-                if(null != image) {
-                    if((null != category) && category.startsWith("Pieces/")) {
```
Comment (giorgiga) [inline MekHQ/src/mekhq/gui/dialog/ImageChoiceDialog.java]: category can't be null here (the if above short-circuits the method if it's null)

File: MekHQ/src/mekhq/gui/view/AtBScenarioViewPanel.java [hunk 12]
```diff
@@ -771,68 +768,13 @@ protected Icon getIcon(Object node) {
         }
 
         protected Icon getIconFrom(UnitStub unit) {
```
Comment (giorgiga) [inline MekHQ/src/mekhq/gui/view/AtBScenarioViewPanel.java]: So many copies of this block of code! (including the comment with the false claim of returning `null`)

```
$ grep -r '// Return a null if the player has selected no portrait file.' src/
src//mekhq/gui/view/ForceViewPanel.java:        // Return a null if the player has selected no portrait file.
src//mekhq/gui/view/ForceViewPanel.java:        // Return a null if the player has selected no portrait file.
src//mekhq/gui/view/ScenarioViewPanel.java:            // Return a null if the player has selected no portrait file.
src//mekhq/gui/view/ScenarioViewPanel.java:            // Return a null if the player has selected no portrait file.
src//mekhq/gui/view/AtBScenarioViewPanel.java:            // Return a null if the player has selected no portrait file.
src//mekhq/gui/view/AtBScenarioViewPanel.java:            // Return a null if the player has selected no portrait file.
src//mekhq/gui/view/PersonViewPanel.java:        // Return a null if the player has selected no portrait file.
src//mekhq/gui/BasicInfo.java:            // Return a null if the player has selected no portrait file.
src//mekhq/gui/BasicInfo.java:            // Return a null if the player has selected no portrait file.
```

File: MekHQ/src/mekhq/util/CamoSpec.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.awt.Color;
+import java.util.Optional;
+
+import megamek.client.ui.swing.util.PlayerColors;
+
+/**
+ * A simple, immutable class holding information on a camo.
+ */
+public class CamoSpec {
```
Comment (giorgiga) [inline MekHQ/src/mekhq/util/CamoSpec.java]: I didn't intend to commit this file! :-(
(will remove as soon as I finish commenting)

File: MekHQ/src/mekhq/util/dom/DomProcessor.java [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+/**
+ * Convenience class to process the children (elements and attributes) of an xml element.
+ * <p>
+ * Unless otherwise noted, all methods respect document-order.
+ */
+public class DomProcessor {
```
Comment (giorgiga) [inline MekHQ/src/mekhq/util/dom/DomProcessor.java]: IMHO this makes parsing sooooo much easier - see for example `ForceStub.generateInstanceFromXML()`.

Let me know what you think of this class: I can remove it and retrofit the few `generateInstanceFromXML` where I've used it or go on and (slowly) spread its use all over the place.

File: MekHQ/unittests/mekhq/util/ImagesTest.java [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+import javax.swing.SwingConstants;
+
+import org.junit.Test;
+
+import mekhq.IconPackage;
+
+@SuppressWarnings({"nls","javadoc"})
+public class ImagesTest {
+
+    public static void main(String[] args) throws Exception {
```
Comment (giorgiga) [inline MekHQ/unittests/mekhq/util/ImagesTest.java]: Not really a unit test, it opens this:

camos

File: MekHQ/src/mekhq/gui/view/UnitViewPanel.java [hunk 16]
```diff
@@ -90,28 +85,28 @@ private void initComponents() {
...
+
+        Image image = Optional.ofNullable(FluffImageHelper.getFluffImage(entity))
+                              .orElseGet(() -> Images.unit(iconPackage, unit, 120, 120));
         Icon icon = null;
-		if(null != image) {
-			if(image.getWidth(lblImage) > 200) {
-                image = image.getScaledInstance(200, -1, Image.SCALE_DEFAULT);               
+        if (null != image) {
+            if (image.getWidth(lblImage) > 120) {
+                image = image.getScaledInstance(120, -1, Image.SCALE_SMOOTH);
```
Comment (giorgiga) [inline MekHQ/src/mekhq/gui/view/UnitViewPanel.java]: 200px was too wide IMHO, so I've made the "fluff" image narrower:  of course I'll revert it to 200 if you like it wider

z

**Comment by NickAragua** [comment]
I gotta say that that third row looks significantly better.

**Comment by giorgiga** [comment]
BTW: I just noticed that MegaMek has been respecting icon colors for some 2 years now (see [relevant commit](https://github.com/MegaMek/megamek/commit/a4b127a8344b86a8aa561c1b6b85b5094994508e)).

**Comment by giorgiga** [comment]
no longer interested in pushing these changes
